codemod_tox/core.bak

```diff
@@ -0,0 +1,236 @@
+                    if c[:com] == prefix[:com]:
+                        new_o = o.removeprefix(c[:com])
+                        if new_o.empty():
+                            new_pieces.append(c[:com])
+                            prefix = prefix[com:]
+                        else:
+                            raise HoistError(f"Leftover nonmatched {prefix!r}")
+                    else:
+                        raise HoistError(f"{o!r} cannot take {prefix!r}")
+        else:
+            raise HoistError("Ran off end of pieces")
+        return self.__class__(new_pieces, self.suffix)
+
+    @classmethod
+    def parse(self, s: str, suffix: str="") -> "ToxEnv":
+        pieces: list[ToxOptions | str] = []
+        for match in TOX_ENV_TOKEN_RE.finditer(s):
+            if match.group("options"):
+                pieces.append(ToxOptions.parse(match.group("options")))
+            elif match.group("literal"):
+                pieces.append(match.group("literal"))
+            else:  # pragma: no cover
+                raise ParseError(f"Bad match {match!r}")
+        return ToxEnv(pieces, suffix)
+
+    def __str__(self) -> str:
+        return "".join(map(str, self.pieces))
+
+
+@dataclass
+class ToxEnvlist(ToxBase):
+    envs: tuple[ToxEnv, ...]
+
+    def all(self) -> Generator[str, None, None]:
+        for e in self.envs:
+            yield from e.all()
+
+    @classmethod
+    def parse(cls, s: str) -> "ToxEnvlist":
+        pieces = []
+        buf = ""
+        default_suffix = "\n"
+
+        for match in TOX_ENV_TOKEN_RE.finditer(s):
+            # This will double-parse but these strings are typically pretty
+            # trivial and we've restricted the character set significantly so
+            # this should be roughly linear
+            if match.group("comma") or match.group("newline"):
+                assert buf
+                if match.group("comma"):
+                    default_suffix = ", "
+                pieces.append(ToxEnv.parse(buf, match.group(0)))
+                buf = ""
+            elif match.group("space"):
+                pass
+            else:
+                buf += match.group(0)
+
+        if buf:
+            pieces.append(ToxEnv.parse(buf, default_suffix))
+
+        return cls(tuple(pieces))
+
+    def __str__(self) -> str:
+        buf = ""
+        default_suffix = "\n"
+        
+        for item, last in _marklast(self.envs):
+            buf += str(item)
+            if not last:
+                if "," in item.suffix:
+                    default_suffix = ", "
+                buf += (item.suffix or default_suffix)
+        return buf
+
+def _marklast(seq: Sequence[T]) -> Generator[tuple[T, bool], None, None]:
+    last_idx = len(seq) - 1
+    for i, x in enumerate(seq):
+        yield x, i==last_idx
+
+def _common_prefix(a: str, b: str) -> str:
+    buf = ""
+    for c1, c2 in zip(a, b):
+        if c1 != c2:
+            break
+        buf += c1
+    return buf
```

Comment:
Aside: Went down the rabbit hole of why we can't use sets for this and then why are sets not order preserving.